### PR TITLE
Issue 855: UPDATE should support table aliasing and FROM clause

### LINKED ISSUE #855: [BigQuery] Found unparsable section: 'UPDATE\n
Hi there!

I am running sqlfluff on a BigQuery update query and the linter throws an `unparsable section` error.

## lint 

```sql
UPDATE
  `my_project.my_table.my_table` AS tttd
SET
  tttd.days=ttu.days,
  tttd.closure=ttu.clousre
FROM
  `{source_table_id}` AS ttu
WHERE
  tttd.po_number=ttu.po_number
;

DROP TABLE `{source_table_id}`
;
```
## parse

```
[0](1, 1, 1)        |file:
[0](1, 1, 1)        |    statement:
[0](1, 1, 1)        |        update_statement:
[0](1, 1, 1)        |            unparsable:                                       !! Expected: 'update_statement'
[0](1, 1, 1)        |                keyword:                                      'UPDATE'
[6](1, 1, 7)        |                newline:                                      '\n'
[7](1, 2, 1)        |                whitespace:                                   '  '
[9](1, 2, 3)        |                raw:                                          '`my_project.my_table.my_table`'
[39](1, 2, 33)      |                whitespace:                                   ' '
[40](1, 2, 34)      |                raw:                                          'AS'
[42](1, 2, 36)      |                whitespace:                                   ' '
[43](1, 2, 37)      |                raw:                                          'tttd'
[47](1, 2, 41)      |                newline:                                      '\n'
[48](1, 3, 1)       |                raw:                                          'SET'
[51](1, 3, 4)       |                newline:                                      '\n'
[52](1, 4, 1)       |                whitespace:                                   '  '
[54](1, 4, 3)       |                raw:                                          'tttd'
[58](1, 4, 7)       |                raw:                                          '.'
[59](1, 4, 8)       |                raw:                                          'days'
[63](1, 4, 12)      |                raw:                                          '='
[64](1, 4, 13)      |                raw:                                          'ttu'
[67](1, 4, 16)      |                raw:                                          '.'
[68](1, 4, 17)      |                raw:                                          'days'
[72](1, 4, 21)      |                comma:                                        ','
[73](1, 4, 22)      |                newline:                                      '\n'
[74](1, 5, 1)       |                whitespace:                                   '  '
[76](1, 5, 3)       |                raw:                                          'tttd'
[80](1, 5, 7)       |                raw:                                          '.'
[81](1, 5, 8)       |                raw:                                          'closure'
[88](1, 5, 15)      |                raw:                                          '='
[89](1, 5, 16)      |                raw:                                          'ttu'
[92](1, 5, 19)      |                raw:                                          '.'
[93](1, 5, 20)      |                raw:                                          'clousre'
[100](1, 5, 27)     |                newline:                                      '\n'
[101](1, 6, 1)      |                raw:                                          'FROM'
[105](1, 6, 5)      |                newline:                                      '\n'
[106](1, 7, 1)      |                whitespace:                                   '  '
[108](1, 7, 3)      |                raw:                                          '`my_table`'
[127](1, 7, 22)     |                whitespace:                                   ' '
[128](1, 7, 23)     |                raw:                                          'AS'
[130](1, 7, 25)     |                whitespace:                                   ' '
[131](1, 7, 26)     |                raw:                                          'ttu'
[134](1, 7, 29)     |                newline:                                      '\n'
[135](1, 8, 1)      |                raw:                                          'WHERE'
[140](1, 8, 6)      |                newline:                                      '\n'
[141](1, 9, 1)      |                whitespace:                                   '  '
[143](1, 9, 3)      |                raw:                                          'tttd'
[147](1, 9, 7)      |                raw:                                          '.'
[148](1, 9, 8)      |                raw:                                          'po_number'
[157](1, 9, 17)     |                raw:                                          '='
[158](1, 9, 18)     |                raw:                                          'ttu'
[161](1, 9, 21)     |                raw:                                          '.'
[162](1, 9, 22)     |                raw:                                          'po_number'
[171](1, 9, 31)     |    newline:                                                  '\n'
[172](1, 10, 1)     |    statement_terminator:                                     ';'
[173](1, 10, 2)     |    newline:                                                  '\n'
[174](1, 11, 1)     |    newline:                                                  '\n'
[175](1, 12, 1)     |    statement:
[175](1, 12, 1)     |        drop_statement:
[175](1, 12, 1)     |            keyword:                                          'DROP'
[179](1, 12, 5)     |            whitespace:                                       ' '
[180](1, 12, 6)     |            keyword:                                          'TABLE'
[185](1, 12, 11)    |            whitespace:                                       ' '
[186](1, 12, 12)    |            table_reference:
[186](1, 12, 12)    |                identifier:                                   '`my_table`'
[205](1, 12, 31)    |    newline:                                                  '\n'
[206](1, 13, 1)     |    statement_terminator:                                     ';'
[207](1, 13, 2)     |    newline:                                                  '\n'

==== parsing violations ====
L:   1 | P:   1 |  PRS | Found unparsable section: 'UPDATE\n `my_project.my_table.my_table`
                       | ...'
```

## Expected Behaviour

No errors

## Observed Behaviour
```
L:   1 | P:   1 |  PRS | Found unparsable section: 'UPDATE\n
                       | `my_project.my_table.my_table` ...'
```

## Steps to Reproduce

```
sqlfluff lint file_name
```

## Version
Include the output of `sqlfluff --version` along with your Python version

## Configuration
```
[sqlfluff]
verbose = 0
dialect = bigquery
templater = python

[sqlfluff:templater:python:context]
source_table_id=my_table
```

### PR DESCRIPTION
**Pull request by barrywhart**
Resolves #855

### CODE REVIEW

File: src/sqlfluff/core/dialects/ansi_keywords.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
 PARTITION
 ORDER
 ROWS
+SET
```
Comment (barrywhart) [inline src/sqlfluff/core/dialects/ansi_keywords.py:22]: Without this change, the parser couldn't handle `UPDATE` statements without aliases. This helps the parser determine the end of the table reference and the start of the "set clause" that updates the data.

File: test/fixtures/parser/ansi/update.yml [hunk 2]
```diff
@@ -3,7 +3,8 @@ file:
     update_statement:
       keyword: UPDATE
       table_reference:
-        identifier: table_name
+        table_reference:
```
Comment (NiallRees) [inline test/fixtures/parser/ansi/update.yml:6]: Nested table_reference? This doesn't look quite right 🤔

**Review by NiallRees — COMMENTED** [review]
I think I'd change the `type` under `AliasedTableReferenceSegment` to be `aliased_table_reference` to be clearer in the parsing structure

File: src/sqlfluff/core/dialects/dialect_ansi.py [hunk 3]
```diff
@@ -2205,8 +2205,9 @@ class UpdateStatementSegment(BaseSegment):
     match_grammar = StartsWith("UPDATE")
     parse_grammar = Sequence(
         "UPDATE",
-        Ref("TableReferenceSegment"),
+        Ref("AliasedTableReferenceSegment"),
```
Comment (NiallRees) [inline src/sqlfluff/core/dialects/dialect_ansi.py:2208]: What's the outcome of changing this line to:
```suggestion
        OneOf(Ref("TableReferenceSegment"), Ref("AliasedTableReferenceSegment"))
```

I think in its current state even un-aliased table reference will be parsed as one. This should solve my comment here https://github.com/sqlfluff/sqlfluff/pull/865/files#r597557373

Comment (barrywhart) [reply src/sqlfluff/core/dialects/dialect_ansi.py:2208]: I'll make that change. I think I was so focused on getting a successful parse, I didn't think about cleanliness. I wonder if we should consider (later, not this PR) changing `AliasedTableReferenceSegment` so the alias is required. The naming is a bit odd, calling it "aliased" when the alias is optional. 🤔

Comment (NiallRees) [reply src/sqlfluff/core/dialects/dialect_ansi.py:2208]: Ah yes I definitely think so. Feel free to do it here or we can tackle separately

Comment (NiallRees) [reply src/sqlfluff/core/dialects/dialect_ansi.py:2208]: I hadn't realised the alias was optional in that segment

**Comment by barrywhart** [comment]
@NiallRees: This is ready for another review -- I made the change you suggested to fix that weird nested `table_reference` I accidentally introduced.

**Review by NiallRees — APPROVED** [review]
I'll take a look at tidying up the table reference grammar

**Pull request merged**